src/main/java/ch/njol/skript/expressions/ExprColorOf.java

```diff
@@ -116,103 +129,155 @@ public Class<?>[] acceptChange(ChangeMode mode) {
-			return CollectionUtils.array(Color.class);
-		if (ItemType.class.isAssignableFrom(returnType))
-			return CollectionUtils.array(Color.class);
-		return null;
+		if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE)
+			return null;
+		return CollectionUtils.array(Color.class);
 	}
 
-	@SuppressWarnings("deprecated")
 	@Override
-	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
-		if (delta == null)
-			return;
-		DyeColor color = ((Color) delta[0]).asDyeColor();
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		DyeColor originalColor = (mode == ChangeMode.DELETE || mode == ChangeMode.RESET) ? null : ((Color) delta[0]).asDyeColor();
 
-		for (Object o : getExpr().getArray(e)) {
-			if (o instanceof Item || o instanceof ItemType) {
-				ItemStack stack = o instanceof Item ? ((Item) o).getItemStack() : ((ItemType) o).getRandom();
+		for (Object obj : getExpr().getArray(event)) {
+			DyeColor color = originalColor; // reset
+			if (obj instanceof Item || obj instanceof ItemType || obj instanceof Slot) {
+				ItemStack stack;
+				if (obj instanceof ItemType) {
+					stack = ((ItemType) obj).getRandom();
+				} else if (obj instanceof Item) {
+					stack = ((Item) obj).getItemStack();
+				} else {
+					stack = ((Slot) obj).getItem();
+				}
 
 				if (stack == null)
 					continue;
 
-				MaterialData data = stack.getData();
-
-				if (!(data instanceof Colorable))
-					continue;
+				org.bukkit.Color bukkitColor = (mode == ChangeMode.DELETE || mode == ChangeMode.RESET) ? null : ((Color) delta[0]).asBukkitColor();
+				ItemMeta meta = stack.getItemMeta();
+				if (meta instanceof LeatherArmorMeta) {
+					LeatherArmorMeta leatherArmorMeta = (LeatherArmorMeta) meta;
+					leatherArmorMeta.setColor(bukkitColor);
+					stack.setItemMeta(leatherArmorMeta);
+				} else if (meta instanceof MapMeta) {
+					MapMeta mapMeta = (MapMeta) meta;
+					mapMeta.setColor(bukkitColor);
+					stack.setItemMeta(mapMeta);
+				} else if (meta instanceof PotionMeta) {
+					PotionMeta potionMeta = (PotionMeta) meta;
+					potionMeta.setColor(bukkitColor);
+					stack.setItemMeta(potionMeta);
+				} else {
+					if (color == null)
+						color = DEFAULT_MATERIAL_COLOR;
+					Material newItem = setMaterialColor(stack.getType(), color);
+					if (newItem == null)
+						continue;
+					ItemMeta oldItemMeta = stack.getItemMeta();
+					stack.setType(newItem);
+					stack.setItemMeta(oldItemMeta);
+					if (obj instanceof ItemType) {
+						ItemType newItemType = new ItemType(newItem);
+						newItemType.setItemMeta(oldItemMeta);
+						((ItemType) obj).setTo(newItemType);
+					} else if (obj instanceof Item) {
+						((Item) obj).setItemStack(stack);
+					} else if (obj instanceof Slot)
+						((Slot) obj).setItem(stack);
+				}
+			} else if (obj instanceof Colorable) {
+				Colorable colorable = getColorable(obj);
+				if (colorable != null) {
+					colorable.setColor(color);
```

Comment:
this worries me
beds throw unsupported exceptions, and nothing else allows null colors so they could throw NPEs.